### PR TITLE
generated agents API changes

### PR DESCRIPTION
**Pull request by benblack769**
* Addresses #441 and #440 
* makes possible_agents optional
* Adds observation_space and action_space methods
* Gives warning when accessing observation_spaces and action_spaces dictionaries
* Gives warning when observation_space and action_space methods are not defined (defaults to accessing observation_spaces and action_spaces dictionaries)
* Created example environments in test folder which have generated agents
* Fixes api_tests

NOT DONE:
* Conversions to and from parallel envs are not yet done
* Parallel API test not yet done
* Warning and error messages not yet reviewed.

### CODE REVIEW

File: pettingzoo/test/api_test.py [hunk 1]
```diff
@@ -38,56 +40,55 @@ def test_observation(observation, observation_0):
...
-        assert isinstance(env.action_spaces[agent], gym.spaces.Space), "Agent space for each agent must extend gym.spaces.Space"
-        if not (isinstance(env.observation_spaces[agent], gym.spaces.Box) or isinstance(env.observation_spaces[agent], gym.spaces.Discrete)):
+        assert isinstance(env.observation_space(agent), gym.spaces.Space), "Observation space for each agent must extend gym.spaces.Space"
+        assert isinstance(env.action_space(agent), gym.spaces.Space), "Agent space for each agent must extend gym.spaces.Space"
+        if not (isinstance(env.observation_space(agent), gym.spaces.Box) or isinstance(env.observation_space(agent), gym.spaces.Discrete)):
             warnings.warn("Observation space for each agent probably should be gym.spaces.box or gym.spaces.discrete")
-        if not (isinstance(env.action_spaces[agent], gym.spaces.Box) or isinstance(env.action_spaces[agent], gym.spaces.Discrete)):
+        if not (isinstance(env.action_space(agent), gym.spaces.Box) or isinstance(env.action_space(agent), gym.spaces.Discrete)):
```
Comment (cclauss) [inline pettingzoo/test/api_test.py:51]: OPTIONAL:
```suggestion
        if not isinstance(env.action_space(agent), (gym.spaces.Box, gym.spaces.Discrete)):
```

File: pettingzoo/test/api_test.py [hunk 2]
```diff
@@ -216,6 +221,13 @@ def progress_report(msg):
             print(msg)
 
     print("Starting API test")
+    if not hasattr(env, 'possible_agents'):
+        warnings.warn(missing_attr_warning.format(name='possible_agents'))
+    if not hasattr(env, 'observation_spaces'):
+        warnings.warn(missing_attr_warning.format(name='observation_spaces'))
+    if not hasattr(env, 'action_spaces'):
+        warnings.warn(missing_attr_warning.format(name='action_spaces'))
```
Comment (cclauss) [inline pettingzoo/test/api_test.py:229]: OPTIONAL:
```suggestion
    for name in ('possible_agents', 'observation_spaces', 'action_spaces'):
        if not hasattr(env, name):
            warnings.warn(missing_attr_warning.format(name=name))
```

File: pettingzoo/test/parallel_test.py [hunk 3]
```diff
@@ -13,12 +22,15 @@ def parallel_api_test(par_env, num_cycles=10):
         assert set(obs.keys()) == (set(par_env.agents))
 
         done = {agent: False for agent in par_env.agents}
-        live_agents = par_env.agents[:]
+        live_agents = set(par_env.agents[:])
         has_finished = set()
         for i in range(num_cycles):
-            actions = {agent: space.sample() for agent, space in par_env.action_spaces.items() if agent in done and not done[agent]}
+            actions = {agent: par_env.action_space(agent).sample() for agent in par_env.agents if agent in done and not done[agent]}
```
Comment (cclauss) [inline pettingzoo/test/parallel_test.py:28]: OPTIONAL: `dict.get()`
```suggestion
            actions = {agent: par_env.action_space(agent).sample() for agent in par_env.agents not done.get(agent, False)}
```

File: pettingzoo/test/parallel_test.py [hunk 4]
```diff
@@ -13,12 +22,15 @@ def parallel_api_test(par_env, num_cycles=10):
...
             obs, rew, done, info = par_env.step(actions)
-            assert par_env.agents == live_agents
+            for agent in par_env.agents:
+                if agent not in live_agents:
```
Comment (cclauss) [inline pettingzoo/test/parallel_test.py:31]:
```suggestion
```
This is a `set` so just add ALL par_env.agents and the set will automatically remove duplicates.

File: pettingzoo/test/save_obs_test.py [hunk 5]
```diff
@@ -5,11 +5,11 @@
...
-        assert isinstance(env.observation_spaces[agent], gym.spaces.Box), "Observations must be Box to save observations as image"
-        assert np.all(np.equal(env.observation_spaces[agent].low, 0)) and np.all(np.equal(env.observation_spaces[agent].high, 255)), "Observations must be 0 to 255 to save as image"
-        assert len(env.observation_spaces[agent].shape) == 3 or len(env.observation_spaces[agent].shape) == 2, "Observations must be 2D or 3D to save as image"
-        if len(env.observation_spaces[agent].shape) == 3:
-            assert env.observation_spaces[agent].shape[2] == 1 or env.observation_spaces[agent].shape[2] == 3, "3D observations can only have 1 or 3 channels to save as an image"
+        assert isinstance(env.observation_space(agent), gym.spaces.Box), "Observations must be Box to save observations as image"
+        assert np.all(np.equal(env.observation_space(agent).low, 0)) and np.all(np.equal(env.observation_space(agent).high, 255)), "Observations must be 0 to 255 to save as image"
+        assert len(env.observation_space(agent).shape) == 3 or len(env.observation_space(agent).shape) == 2, "Observations must be 2D or 3D to save as image"
```
Comment (cclauss) [inline pettingzoo/test/save_obs_test.py:10]:
```suggestion
        assert len(env.observation_space(agent).shape) in (2, 3), "Observations must be 2D or 3D to save as image"
```

File: pettingzoo/test/save_obs_test.py [hunk 6]
```diff
@@ -5,11 +5,11 @@
...
+        if len(env.observation_space(agent).shape) == 3:
+            assert env.observation_space(agent).shape[2] == 1 or env.observation_space(agent).shape[2] == 3, "3D observations can only have 1 or 3 channels to save as an image"
```
Comment (cclauss) [inline pettingzoo/test/save_obs_test.py:12]:
```suggestion
            assert env.observation_space(agent).shape[2] in (1, 3), "3D observations can only have 1 or 3 channels to save as an image"
```

**Comment by benblack769** [comment]
@cclauss Thanks for pointing out these nice python features and style improvements. I'll take a look soon.

**Pull request merged**